cns/restserver/ipam.go

```diff
@@ -478,101 +627,158 @@ func (service *HTTPRestService) MarkExistingIPsAsPendingRelease(pendingIPIDs []s
+			ncMap[ipConfig.NCID] = ipConfig
+			if len(ncMap) == len(desiredIPAddresses) {
+				return podIPInfo, err
 			}
-			err := service.populateIPConfigInfoUntransacted(ipConfig, &podIpInfo)
-			return podIpInfo, err
 		}
 	}
-	return podIpInfo, fmt.Errorf("Requested IP not found in pool")
+
+	// if we were able to get at least one IP but not all of the desired IPs
+	if len(ncMap) > 0 {
+		logger.Printf("[AssignDesiredIPConfigs] Failed to retrieve all desired IPs. Releasing all IPs that were found")
+		for _, ipState := range ncMap { //nolint:gocritic // ignore copy
+			_, err := service.unassignIPConfig(ipState, podInfo)
+			if err != nil {
+				return podIPInfo, fmt.Errorf("[AssignDesiredIPConfigs] failed to mark IPConfig [%+v] back to Available. err: %w", ipState, err)
+			}
+		}
+	}
+	//nolint:goerr113 // return error
+	return podIPInfo, fmt.Errorf("Not all requested ips were found/available in the pool")
 }
 
-func (service *HTTPRestService) AssignAnyAvailableIPConfig(podInfo cns.PodInfo) (cns.PodIpInfo, error) {
+// Assigns an IP from each NC on the NNC
+func (service *HTTPRestService) AssignAvailableIPConfigs(podInfo cns.PodInfo) ([]cns.PodIpInfo, error) {
 	service.Lock()
 	defer service.Unlock()
+	// Creates a slice of PoPodIpInfo with the size of number of NCs
+	podIPInfo := make([]cns.PodIpInfo, len(service.state.ContainerStatus))
+	// This map is used to store whether or not we have found an available IP from an NC when looping through the pool
+	ncMap := make(map[string]cns.IPConfigurationStatus)
 
 	for _, ipState := range service.PodIPConfigState {
-		if ipState.GetState() == types.Available {
+		_, found := ncMap[ipState.NCID]
+		// Checks if the current IP is available and if we haven't already found an IP from that NC
+		if !found && ipState.GetState() == types.Available {
 			if err := service.assignIPConfig(ipState, podInfo); err != nil {
-				return cns.PodIpInfo{}, err
+				logger.Errorf(err.Error())
+				break
 			}
 
-			podIPInfo := cns.PodIpInfo{}
-			if err := service.populateIPConfigInfoUntransacted(ipState, &podIPInfo); err != nil {
-				return cns.PodIpInfo{}, err
+			if err := service.populateIPConfigInfoUntransacted(ipState, &podIPInfo[len(ncMap)]); err != nil {
+				logger.Errorf(err.Error())
+				break
+			}
+			ncMap[ipState.NCID] = ipState
+			// return once we have found one IP per NC
+			if len(ncMap) == len(service.state.ContainerStatus) {
+				return podIPInfo, nil
 			}
+		}
+	}
 
-			return podIPInfo, nil
+	// if we were able to find at least one IP but not enough
+	if len(ncMap) > 0 {
+		logger.Printf("[AssignAvailableIPConfigs] Failed to retrieve enough IPs. Releasing all IPs that were found")
+		for _, ipState := range ncMap { //nolint:gocritic // ignore copy
+			_, err := service.unassignIPConfig(ipState, podInfo)
+			if err != nil {
+				return podIPInfo, fmt.Errorf("[AssignAvailableIPConfigs] failed to mark IPConfig [%+v] back to Available. err: %w", ipState, err)
+			}
 		}
 	}
```

Comment:
We have a lock on the service during this method, so I wonder if it would be better to lazily reserve the IPs rather than opportunistically reserving them and then maybe needing to unreserve them immediately again if we do not find as many as we need.
We could rearrange this method: in the (above) first iteration, we attempt to identify available IPs from each NC, and we put them in the NC map. Then, **once we have found enough IPs to satisfy the request**, we iterate the NC map and reserve those IPs via the 
```go
service.assignIPConfig
``` 
and 
```go
service.populateIPConfigInfoUntransacted
```
calls. If we fail to assign at any point, we can immediately attempt to unassign the whole list, like we already do here. But this will decrease churn and lock contention in scenarios where there are not enough IPs from all NCs to satisfy the request, and the code will flow better.

Reply:
consider this for other methods in this file, also